modem: chat: Add runtime API to modem chat matches

bjarki-andreasen opened the pull request:
Adds dedicated API for manipulating `struct modem_chat_match` instances at runtime, to improve the currently working, but rather unsafe and wordy solution:
```
data->dynamic_match.match = data->dynamic_match_buf;
	data->dynamic_match.separators = data->dynamic_separators_buf;
	data->dynamic_match.separators_size = sizeof(data->dynamic_separators_buf);
	data->dynamic_match.wildcards = false;
	data->dynamic_match.partial = false;

	data->dynamic_script_chat.request = data->dynamic_request_buf;
	data->dynamic_script_chat.response_matches = &data->dynamic_match;
	data->dynamic_script_chat.response_matches_size = 1;
	data->dynamic_script_chat.timeout = 0;

	data->dynamic_script.name = "pair";
	data->dynamic_script.script_chats = &data->dynamic_script_chat;
	data->dynamic_script.script_chats_size = 1;
	data->dynamic_script.abort_matches = NULL;
	data->dynamic_script.abort_matches_size = 0;
	data->dynamic_script.callback = NULL;
	data->dynamic_script.timeout = 10;
```
turning it into:
```
	modem_chat_match_init(&data->pair_match);
	modem_chat_match_set_separators(&data->pair_match, ",*");

	modem_chat_script_chat_init(&data->pair_script_chat);
	modem_chat_script_chat_set_response_matches(&data->pair_script_chat, &data->pair_match, 1);

	modem_chat_script_init(&data->pair_script);
	modem_chat_script_set_name(&data->pair_script, "pair");
	modem_chat_script_set_script_chats(&data->pair_script, &data->pair_script_chat, 1);
	modem_chat_script_set_abort_matches(&data->pair_script, &data->pair_match, 1);
	modem_chat_script_set_timeout(&data->pair_script, 10);
```
This is way safer since the members are not directly interacted with, member values are validated when set. A notable safety improvement is updating the size of a modem_chat to match the string length, which was manually done before this PR like this:
```
	/* update buffer first */
	ret = gnss_nmea0183_snprintk(data->pair_request_buf, sizeof(data->pair_request_buf),
				     "PAIR752,%u,%u", pps_mode, config->pps_pulse_width);
	if (ret < 0) {
		return ret;
	}

	/* manually set request size to match strlen of buffer (not checking if length exceeds UINT16_MAX) */
	data->pair_script_chat.request_size = ret;
```
After this PR, the buffer is first updated, then set using the new runtime API
```
	/* update buffer first */
	ret = gnss_nmea0183_snprintk(data->pair_request_buf, sizeof(data->pair_request_buf),
				     "PAIR752,%u,%u", pps_mode, config->pps_pulse_width);
	if (ret < 0) {
		return ret;
	}

	/* set buffer using runtime API */
	ret = modem_chat_script_chat_set_request(&data->pair_script_chat, data->pair_request_buf);
	if (ret < 0) {
		return ret;
	}
```

bjarki-andreasen committed:
modem: chat: Add runtime API to modem chat matches

Add APIs for changing the contents of modem chat matches
safely at runtime.

This allows for reusing a single modem_chat_match at the cost
of placing the match and its buffers in RAM.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

rerickson1 commented on subsys/modem/modem_chat.c:
```diff
@@ -835,3 +836,104 @@ void modem_chat_release(struct modem_chat *chat)
+				   uint8_t match_buf_size, uint8_t *separators_buf,
+				   uint8_t separators_buf_size)
+{
+	__ASSERT(runtime != NULL, "runtime must be provided");
+	__ASSERT(match_buf != NULL, "match_buf must be provided");
+	__ASSERT(match_buf_size > 1, "match_buf_size must more than ");
+	__ASSERT(separators_buf != NULL, "separators_buf must be provided");
+	__ASSERT(separators_buf_size > 0, "separators_buf_size must more than 0");
```
Will these only assert if assert is enabled (kconfig) or will they always assert?

bjarki-trackunit replied:
Only if enabled, the idea being that you would disable the asserts when building production code to save space/overhead :)

rerickson1 replied:
Got it. Wouldn't we want this stuff checked without asserts?

bjarki-trackunit replied:
I only place these asserts in init functions which are only run once, usually in the device drivers init function.This is to fail fast since there is no recovery action that can be made if the init fails.
This pattern turns this
```
int ret = modem_chat_match_init(&foo, &bar);
if (ret < 0) {
        return ret;
}
```
into the simpler
```
modem_chat_match_init(&foo, &bar);
```
which simply explodes if not initialized properly. Then, for production code, the asserts are disabled, which is safe as the variables passed to the init functions don't change at runtime, they are statically allocated buffers and const configurations :)

bjarki-andreasen committed:
modem: chat: Add runtime API to modem_chat_script_chat

Add APIs allowing for modifying the modem_chat_script_chat
safely at runtime.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
modem: chat: Add runtime APIs for chat scripts

Add runtime APIs for chat scripts.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
drivers: gnss: LCX6G use modem chat runtime API

Update the Quectel LCX6G device driver to use the modem chat
runtime API to modify the dynamic pair chat script.

This change makes the driver safer and more readible by
ensuring safe modification of the members of the modem chat
structures.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-trackunit commented:
Ready for review, it has changed quite a bit from the first round, so please re-review with fresh eyes

tomi-font commented on include/zephyr/modem/chat.h:
```diff
@@ -376,6 +376,156 @@ void modem_chat_script_abort(struct modem_chat *chat);
+ * @param chat_match Modem chat match instance
+ * @param wildcards Wildcards flag to set
+ */
+void modem_chat_match_set_wildcards(struct modem_chat_match *chat_match, bool wildcards);
```
Not introduced by this PR, but I think that `wildcards` would benefit from an improved naming, e.g. `uses_wildcards`, `wildcard_enabled`... Because currently the naming could mean that what you are dealing with is the actual wildcards (although it becomes obvious that it's not the case when you have a closer look).
Looking at it, why not also declaring it (and `partial`) as bool instead of `uint8_t : 1`? :slightly_smiling_face:

bjarki-trackunit replied:
Agreed, changed the signature to:
```
void modem_chat_match_enable_wildcards(struct modem_chat_match *chat_match, bool enable);
```
this should read correctly :) Also changed the members  to `bool`

tomi-font commented on include/zephyr/modem/chat.h:
```diff
@@ -376,6 +376,156 @@ void modem_chat_script_abort(struct modem_chat *chat);
+ * @param script_chat Modem chat script chat instance
+ * @param request Request to set
+ * @note The lifetime of request must match or exceed the lifetime of script_chat
+ * @warning Always call this API after Request is modified
```
```suggestion
 * @warning Always call this API after request is modified
```

bjarki-trackunit replied:
Done

tomi-font commented on include/zephyr/modem/chat.h:
```diff
@@ -376,6 +376,156 @@ void modem_chat_script_abort(struct modem_chat *chat);
+ * @note The lifetime of response_matches must match or exceed the lifetime of script_chat
+ *
+ * @retval 0 if successful
+ * @retval -EINVAL if response_matches_size > 0 and response_matches == NULL
```
This is not the whole story here. `-EINVAL` is returned also if `reponses_matches != NULL && response_matches_size == 0`.
To me this is too much implementation detail exposed in the header file, that also has a risk of falling out of sync, just as is done in other places ([here](https://github.com/zephyrproject-rtos/zephyr/pull/69501/files#diff-be5929fe577613cc402559752fe245ca1dd5bf6148588a10a6d24ec0d3e7e384R405), [here](https://github.com/zephyrproject-rtos/zephyr/pull/69501/files#diff-be5929fe577613cc402559752fe245ca1dd5bf6148588a10a6d24ec0d3e7e384R508)).
How about instead stating something a tad more generic?

bjarki-trackunit replied:
Generalized to
```
 * @retval 0 if successful, negative errno code otherwise
```

tomi-font replied:
Nice!

tomi-font commented on include/zephyr/modem/chat.h:
```diff
@@ -376,6 +376,156 @@ void modem_chat_script_abort(struct modem_chat *chat);
+ * @param timeout Timeout in milliseconds
+ */
+void modem_chat_script_chat_set_timeout(struct modem_chat_script_chat *script_chat,
+					uint16_t timeout);
```
```suggestion
					uint16_t timeout_ms);
```
for extra clarity

bjarki-trackunit replied:
Done

tomi-font commented on include/zephyr/modem/chat.h:
```diff
@@ -376,6 +376,156 @@ void modem_chat_script_abort(struct modem_chat *chat);
+				       uint16_t response_matches_size);
+
+/**
+ * @brief Set modem chat script chat matches
```
Ahem? :slightly_smiling_face: 
I am questioning a bit the value of these comments...

bjarki-trackunit replied:
Most of them are quite self explanatory from the name of the function, I add the brief for a bit more grammatically correct documentation, I agree it may be a bit overkill/redundant, but it

bjarki-trackunit replied:
oh, got distracted :D ... but it is harmless :)

tomi-font commented on include/zephyr/modem/chat.h:
```diff
@@ -376,6 +376,156 @@ void modem_chat_script_abort(struct modem_chat *chat);
+ * @retval 0 if successful
+ * @retval -EINVAL if abort_matches_size > 0 and abort_matches == NULL
+ */
+int modem_chat_script_set_matches(struct modem_chat_script *script,
```
```suggestion
int modem_chat_script_set_abort_matches(struct modem_chat_script *script,
```
maybe?
(And this makes me question the naming of `modem_chat_script_chat_set_matches()`, whether to add `response` to it.)

bjarki-trackunit replied:
Added response and abort to the names, it is definitely better, the names are just getting l o n g :)

tomi-font commented on include/zephyr/modem/chat.h:
```diff
@@ -376,6 +376,156 @@ void modem_chat_script_abort(struct modem_chat *chat);
+ * @param script Modem chat script instance
+ * @param timeout Timeout in seconds
+ */
+void modem_chat_script_set_timeout(struct modem_chat_script *script, uint32_t timeout);
```
```suggestion
void modem_chat_script_set_timeout(struct modem_chat_script *script, uint32_t timeout_s);
```

bjarki-trackunit replied:
Done

tomi-font reviewed (changes requested):
Looks good to me, minus minor remarks.

tomi-font commented on include/zephyr/modem/chat.h:
```diff
@@ -45,9 +45,9 @@ struct modem_chat_match {
+	bool wildcards : 1;
 	/** Set if script shall not continue to next step in case of match */
-	uint8_t partial : 1;
+	bool partial : 1;
```
No need for the ` : 1` part anymore?

bjarki-trackunit replied:
The size of a bool is 1 byte, so we save a byte by using a bitfield :)

bbilas replied:
I wouldn't use a bitfield here as it may cause other alignment/filling issues and performance may suffer (we may burn some cycles due to masking and shifting). Is it really worth saving just a byte in that case?

bjarki-trackunit replied:
bitfield has been removed :)
